Improve assert message handling, particularly on macOS

rcombs opened the pull request:
- Add a test case for assertion failures (separate from the direct OPENSSL_die call in the existing abort test)
- Add the asserting function's name to assertion-failure error messages
- Add the assertion-failure error message to crash dumps on macOS and iOS

This improves debug output on assertion failures, particularly when an Apple crash dump is available but stderr is not.

##### Checklist
- [X] tests are added or updated

rcombs commented:
Submitted a CLA under my current email address.

mattcaswell commented:
Close/reopen to kick CLA bot

mattcaswell commented on crypto/cryptlib.c:
```diff
@@ -396,15 +396,34 @@ void OPENSSL_showfatal(const char *fmta, ...)
+#   if (defined(__MAC_10_11)  && __MAC_OS_X_VERSION_MIN_REQUIRED  >= __MAC_10_11) || \
+       (defined(__IPHONE_9_0) && __IPHONE_OS_VERSION_MIN_REQUIRED >= __IPHONE_9_0)
+#    define HAVE_OS_CRASH
+extern void _os_crash(const char *);
```
This appears to be an undocumented API. I'm not sure it is wise to use it?

rcombs replied:
Hmmm, we could weak-link to it (which would mean we could check at runtime if the symbol is present or not). Alternately, we could use `__assert_rtn`, which is also undocumented but is called from the `assert()` macro and thus has frozen ABI, but that requires more structured input, and logs to stderr as well; see e.g.:
```
Assertion failed: (!"Voluntary assertion failure"), function main, file test/asserttest.c, line 14.
```
I suppose we could add `OpenSSL internal error: ` before the message with an asprintf?

mattcaswell commented on crypto/cryptlib.c:
```diff
@@ -413,10 +432,13 @@ int OPENSSL_isservice(void)
 #endif
 
-void OPENSSL_die(const char *message, const char *file, int line)
+void OPENSSL_die2(const char *message, const char *file, int line, const char *func)
```
Is adding a new public API function justified for this? If we were to add it, it should be called `OPENSSL_die_ex`, and should come with documentation in the pod files.

rcombs replied:
`OPENSSL_die` is already public-but-undocumented so I didn't see it as a major change. Not entirely sure _why_ it's public tbh (along with `OPENSSL_assert`); should those be deprecated from the public API?

mattcaswell replied:
Perhaps they should be deprecated. I don't see a need to add any *more* functions like this. I'm also not really sure why we need it anyway. It seems somewhat tangential to the main point of this PR.

richsalz replied:
Perhaps more importantly, OpenSSL *removed all internal function codes* from its error stack.  Adding this back seems wrong.

levitte replied:
Er, `func` here has nothing to do with the internal function codes, and everything to do with compiler generated function *names*, which we do use instead of the old function codes when available.

I have told you this before, @richsalz

richsalz replied:
But functiona names are not currently used in the assertion stuff.

levitte replied:
And? That is of course a choice, but I'd rather not factor in facts that have no relevance here.

richsalz replied:
We disagree.

It doesn't seem right to add a new undocumented function, tho.

kaduk replied:
Isn't it against policy to add new undocumented functions?

rcombs replied:
I'd agree that this should be internal, I'm just not sure how to deal with the existing public macro (nor how to deprecate things in general).

mattcaswell commented on test/recipes/01-test_assert.t:
```diff
@@ -0,0 +1,16 @@
+#! /usr/bin/env perl
+# Copyright 2015-2016 The OpenSSL Project Authors. All Rights Reserved.
```
2020

mattcaswell commented on crypto/cryptlib.c:
```diff
@@ -396,15 +396,34 @@ void OPENSSL_showfatal(const char *fmta, ...)
+# ifdef __APPLE__
+#  include <Availability.h>
+#   if (defined(__MAC_10_11)  && __MAC_OS_X_VERSION_MIN_REQUIRED  >= __MAC_10_11) || \
+       (defined(__IPHONE_9_0) && __IPHONE_OS_VERSION_MIN_REQUIRED >= __IPHONE_9_0)
```
We've had problems in the past with some compilers falling over with expressions of this form:

````
#if defined(MYMACRO) && MYMACRO > 0
````

The problem being that if the macro is not defined they do not short circuit and still attempt to evaluate the second part of the expression.

rcombs replied:
Removed these ifdefs in favor of a runtime check for a weak symbol.

mattcaswell commented on crypto/cryptlib.c:
```diff
@@ -396,15 +396,34 @@ void OPENSSL_showfatal(const char *fmta, ...)
+
+#ifdef HAVE_OS_CRASH
+    va_list ap2;
+    va_copy(ap2, ap);
```
We don't allow mixed declarations and code. Declarations should always be at the beginning of the block.

mattcaswell commented on crypto/cryptlib.c:
```diff
@@ -413,10 +432,13 @@ int OPENSSL_isservice(void)
 {
-    OPENSSL_showfatal("%s:%d: OpenSSL internal error: %s\n",
-                      file, line, message);
+    if (!func)
```
Use `func != NULL`

rcombs commented:
Addressed code comments and switched to a runtime weak-symbol check; still needs further discussion on the added function (which should probably be internal but the assert macro being public is awkward) and the undocumented API usage.

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 30 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

rcombs commented:
I think this needs some more discussion to figure out what should happen to `OPENSSL_die` (and any successor function).

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 30 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 61 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 92 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 123 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 154 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).

openssl-machine commented:
This PR is waiting for the creator to make requested changes but it has not been updated for 185 days.  If you have made changes or commented to the reviewer please make sure you re-request a review (see icon in the 'reviewers' section).
